src/shared/cgroup-setup.c

```diff
@@ -424,26 +424,41 @@ static int cg_set_access_one(
                 const char *controller,
                 const char *path,
                 uid_t uid,
-                gid_t gid) {
+                gid_t gid,
+                bool top) {
 
         struct Attribute {
                 const char *name;
                 bool fatal;
+                bool in_top;
         };
 
         /* cgroup v1, aka legacy/non-unified */
         static const struct Attribute legacy_attributes[] = {
-                { "cgroup.procs",           true  },
-                { "tasks",                  false },
-                { "cgroup.clone_children",  false },
+                { "cgroup.procs",           true,  true  },
+                { "cgroup.clone_children",  false, true  },
+                { "tasks",                  false, true  },
+                { "notify_on_release",      false, false },
                 {},
         };
 
         /* cgroup v2, aka unified */
         static const struct Attribute unified_attributes[] = {
-                { "cgroup.procs",           true  },
-                { "cgroup.subtree_control", true  },
-                { "cgroup.threads",         false },
+                { "cgroup.procs",           true,  true },
+                { "cgroup.subtree_control", true,  true },
+                { "cgroup.threads",         false, true },
+                { "cgroup.controllers",     false, false},
+                { "cgroup.events",          false, false},
+                { "cgroup.freeze",          false, false},
+                { "cgroup.kill",            false, false},
+                { "cgroup.max.depth",       false, false},
+                { "cgroup.max.descendants", false, false},
+                { "cgroup.stat",            false, false},
+                { "cgroup.type",            false, false},
+                { "cpu.pressure",           false, false},
+                { "cpu.stat",               false, false},
+                { "io.pressure",            false, false},
+                { "memory.pressure",        false, false},
```

Comment:
hmm, shouldn't we just chown everything in the subcgroup? i.e. just use recurse_dir() and fchwon() everything recursively?

Reply:
I'm not proud of this part but complete recursion won't make it because there are files that are "owned" by the delegator but are placed in the delegatee's subtree. E.g. delegatee shouldn't tune it's `memory.max` or `cgroup.max.depth` (the former example is illustrative only, subcgroup creation happens before controller enablement).
However, it's perfectly fine to modify these _any_ of the knobs deeper in the hierarchy by the delegatee.
To make the cause not so simple, _some_ 1st level knobs must actually be permitted to the delegatee (`cgroup.procs` and `cgroup.subtree_control`).
That's why this function ended up with listing all possible cgroup (core) attributes :-/

Reply:
i don't follow?

if we define a subcgroup "foo" then everything inside of "foo" should just be chowned to user of service. Why wouldn't we?